Use MbedCRC for LittleFS

deepikabhavnani opened the pull request:
### Description

916dec9fe4d7f7fc376c5979550b17a247d7173b : CRC used in LittleFS is Reversed ANSI, hence new polynomial added. Reversed polynomials perform shift in reverse direction of standard polynomial, and we do not have option to notify reverse shift to hardware. Hence this option is available in software only.

### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

~~Dependent on https://github.com/ARMmbed/mbed-os/pull/7793~~

deepikabhavnani commented:
Preceding PR merged, hence rebased

Commit:
Use MbedCRC for LittleFS (0xEDB88320)

CRC used in LittleFS is Reversed ANSI, hence new polynomial added.
Reversed polynomials perform shift in reverse direction of standard
polynomial, and we do not have option to notify reverse shift to hardware.
Hence this option is available in software only.

geky commented on features/filesystem/littlefs/littlefs/lfs_util.c:
```diff
@@ -6,11 +6,19 @@
 // Only compile if user does not provide custom config
 #ifndef LFS_CONFIG
 
+#ifdef __MBED__
```
nit: Could this be moved to the header file? And just make the crc function declared LittleFileSystem.cpp named "lfs_crc"?

geky commented:
Looks good :+1: Though need to double check that the resulting disk image is portable.

Commit:
Use the same API name to add mbed CRC: lfs_crc

0xc0170 commented:
@deepikabhavnani Is this ready for review or still needs work?

deepikabhavnani commented:
@0xc0170 - Its ready for review

0xc0170 commented on features/filesystem/littlefs/littlefs/lfs_util.c:
```diff
@@ -6,11 +6,10 @@
-
-
-// Software CRC implementation with small lookup table
+#ifndef __MBED__
```
shouldn't we rather provide LFS_CONFIG than changing directly littlefs (another update might replace this?)

deepikabhavnani replied:
@0xc0170 - Yes we can but I assumed LFS_CONFIG is apart from mbed-os changes. https://github.com/ARMmbed/mbed-os/blob/6bb2f500cddcfb707920ff866b571074cc8d0071/features/filesystem/littlefs/littlefs/lfs_util.h#L47

geky replied:
LFS_CONFIG was added after we first integrated LittleFS into Mbed OS, so that's why lfs_util.c/h have been modified directly.

git subtrees handle this surprisingly gracefully, so moving to LFS_CONFIG has been low priority.

deepikabhavnani replied:
As part of this PR, shall we update "lfs_util.c/h" or add separate files?

geky replied:
Up to you, adding separate files can be a separate PR

deepikabhavnani replied:
> Up to you, adding separate files can be a separate PR

Will stay with this for now. Thanks

geky commented:
This has a risk of effecting LittleFS structures on disk. @cmonr Could I request we hold off on this until 5.10?

cmonr commented:
> This has a risk of effecting LittleFS structures on disk. @cmonr Could I request we hold off on this until 5.10?

This is why I like to tag PRs as soon as possible :)

deepikabhavnani commented:
Code freeze for 5.10 is before 5.9.7 and this PR is on top of other 3 CRC PR’s. Please tag all 4 to same release

cmonr commented:
@geky  @0xc0170 Good with this PR?

0xc0170 commented:
@geky please finalize the review

geky commented:
Sorry for delay, I wanted to verify that this did not break portability. It looks good to me :+1: 

@deepikabhavnani, thanks for putting this together

NirSonnenschein commented:
This will proabaly need a rebase.

cmonr commented:
/morph build

cmonr commented:
@deepikabhavnani @geky Fyi, I rebased the PR. Let us know if anything looks amis.

0xc0170 commented:
2 test failures related to the filesystem, please review

geky committed:
CRC: Removed assertion on NULL buffer

When length is zero, the buffer is not accessed. The CRC functions are used
inside several other layers where a 0-length buffer may have meanings in
different contexts and being able to pass 0-length NULL buffers to CRC as a
noop simplifies the code.

adbridge commented:
/morph build

cmonr commented:
Marking this for RC2.

cmonr commented:
/morph export-build

cmonr commented:
/morph test
